packages/visual-editor/src/utils/filterComponents.tsx

```diff
@@ -0,0 +1,69 @@
+import { DefaultComponentProps, type Config } from "@measured/puck";
+
+const gatedLayoutComponents: string[] = ["CustomCodeSection", "GridSection"];
+
+/**
+ * Filters out gated layout components from a configuration object unless explicitly allowed.
+ *
+ * @param config - The original configuration object.
+ * @param additionalLayoutComponents - An optional list of gated component names to retain.
+ * @returns The new configuration object with gated components that are not present in `additionalLayoutComponents` removed.
+ */
+export const filterComponentsFromConfig = <T extends DefaultComponentProps>(
+  config: Config<T>,
+  additionalLayoutComponents?: string[]
+): Config<T> => {
+  // Filter components object
+  const filteredComponents = Object.fromEntries(
+    Object.entries(config.components).filter(
+      ([key]) =>
+        !gatedLayoutComponents.includes(key) ||
+        additionalLayoutComponents?.includes(key)
+    )
+  ) as Config<T>["components"];
+
+  // Filter categories by removing gated components that are not present in the additional components list from their components arrays
+  const filteredCategories = Object.fromEntries(
+    Object.entries(config.categories || {}).map(([categoryKey, category]) => [
+      categoryKey,
+      {
+        ...category,
+        components: (category.components || []).filter(
+          (componentName) =>
+            !gatedLayoutComponents.includes(componentName as string) ||
+            additionalLayoutComponents?.includes(componentName as string)
+        ),
+      },
+    ])
+  );
+
+  return {
+    ...config,
+    components: filteredComponents,
+    categories: filteredCategories,
+  };
+};
+
+/**
+ * Filters gated layout components from a specific configuration in the registry.
+ *
+ * @param registry - The registry to be filtered.
+ * @param registryKey - The key in the registry whose associated configuration should be filtered.
+ * @param additionalLayoutComponents - An optional list of gated component names to retain.
+ * @returns The new registry with the filtered configuration under the specified key.
+ */
+export const filterComponentsFromRegistry = (
```

Comment:
I feel like this isn't really necessary, at least to be exported. The user can use filterComponentsFromConfig directly since they're already working with the correct config for the template.